### PR TITLE
[DataGrid] EF Core Adapter - A second operation was started

### PR DESCRIPTION
**Pull request by miguelhasse**
# Pull Request

## 📖 Description

DataGrid with Paginator using IAsyncQueryExecutor for Entity Framework fails with the following error:
"A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext."

### 🎫 Issues

[fix: A second operation was started in DataGrid with EF Core #801](https://github.com/microsoft/fluentui-blazor/issues/801)

## 📑 Test Plan

Create a Blazor Server project that uses EF to connect a DataGrid component to a large database table, using IQueryable and pagination. Then quickly browse through pages to check there is no "second operation was started" exception. Ideally this should be done using integration tests using something like [Testcontainers](https://dotnet.testcontainers.org).

### General

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fluentui-blazor/blob/master/docs/contributing.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have added [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for my new component
- [x] I have modified an existing component
- [ ] I have validated the [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for an existing component

### CODE REVIEW

**Comment by miguelhasse** [comment]
> microsoft

@microsoft-github-policy-service agree

**Comment by bryansjenks** [comment]
I'll just note that we have this problem all over and if this fix addresses it; that would be really helpful.
Thanks everyone for your hard work on this project!
-Bryan

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 1]
```diff
@@ -471,18 +475,26 @@ public Task SortByColumnAsync(ColumnBase<TGridItem> column, SortDirection direct
...
 
         return column is not null ? SortByColumnAsync(column, direction) : Task.CompletedTask;
     }
 
+    /// <summary>
+    /// Removes the grid's sort on double click for the currently sorted column if it's not a default sort column.
+    /// </summary>
+    /// <returns>A <see cref="Task"/> representing the completion of the operation.</returns>
+    public Task RemoveSortByColumnAsync() => (_sortByColumn != null) ? RemoveSortByColumnAsync(_sortByColumn) : Task.CompletedTask;
+
```
Comment (vnbaaij) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:491]: Although a nice addition, this is not relevant for the PR subject. Please remove and create a new PR for this enhancement

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 2]
```diff
@@ -518,11 +531,33 @@ public Task ShowColumnOptionsAsync(ColumnBase<TGridItem> column)
...
+    /// <summary>
+    /// Displays the <see cref="ColumnBase{TGridItem}.ColumnOptions"/> UI for the specified column <paramref name="index"/>,
+    /// closing any other column options UI that was previously displayed. If the index is out of range, nothing happens.
+    /// </summary>
+    /// <returns>A <see cref="Task"/> representing the completion of the operation.</returns>
+    public Task ShowColumnOptionsAsync(int index)
+    {
+        return (index >= 0 && index < _columns.Count) ? ShowColumnOptionsAsync(_columns[index]) :  Task.CompletedTask;
+    }
+
```
Comment (vnbaaij) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:554]: Although a nice addition, this is not relevant for the PR subject. Please remove and create a new PR for this enhancement

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 3]
```diff
@@ -541,10 +576,7 @@ public void SetLoadingState(bool loading)
     /// (either <see cref="Items"/> or <see cref="ItemsProvider"/>).
     /// </summary>
     /// <returns>A <see cref="Task"/> that represents the completion of the operation.</returns>
-    public async Task RefreshDataAsync()
-    {
-        await RefreshDataCoreAsync();
-    }
+    public Task RefreshDataAsync() => RefreshDataCoreAsync();
```
Comment (vnbaaij) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:579]: Using expression body is fine, but why remove the async / await? Think we should leave that in.

**Review by vnbaaij — CHANGES_REQUESTED** [review]

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs (hunk 3 above)
Comment (miguelhasse) [reply src/Core/Components/DataGrid/FluentDataGrid.razor.cs:579]: Async/await is unnecessary here and removing it doesn't change behavior but prevents the generation of an unnecessary underlying state machine by the compiler. Any way, I will remove it from this PR.

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs (hunk 1 above)
Comment (miguelhasse) [reply src/Core/Components/DataGrid/FluentDataGrid.razor.cs:491]: Sure, I'll move this change into another PR as you sugested. (I will only be able to do this next Monday)

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs (hunk 2 above)
Comment (miguelhasse) [reply src/Core/Components/DataGrid/FluentDataGrid.razor.cs:554]: Sure, I'll move this change into another PR as you sugested. (I will only be able to do this next Monday)

**Review by dvoituron — APPROVED** [review]

**Review by vnbaaij — APPROVED** [review]

**Comment by miguelhasse** [comment]
@bryansjenks I've provided a working sample in the following branch https://github.com/miguelhasse/fluentui-blazor/tree/adventureworks-server-sample/examples/AdventureWorks

**Comment by miguelhasse** [comment]
@bryansjenks I've provided a working sample in the following branch https://github.com/miguelhasse/fluentui-blazor/tree/adventureworks-server-sample/examples/AdventureWorks

**Pull request merged**